components/encryption/src/master_key/kms.rs

```diff
@@ -0,0 +1,306 @@
+            self.client
+                .generate_data_key(generate_request.clone())
+                .map_err(|e| Error::Other(e.into()))
+        });
+        let ciphertext_key = generate_response.ciphertext_blob.unwrap().as_ref().to_vec();
+        let plaintext_key = generate_response.plaintext.unwrap().as_ref().to_vec();
+        Ok((ciphertext_key, plaintext_key))
+    }
+}
+
+fn retry<T, U, F>(runtime: &mut Runtime, mut func: F) -> T
+where
+    F: FnMut() -> U,
+    U: Future<Output = Result<T>> + std::marker::Unpin,
+{
+    let retry_limit = 6;
+    let timeout_duration = Duration::from_secs(10);
+    for _ in 0..retry_limit {
+        let fut = func();
+
+        match runtime.block_on(async move {
+            let timeout = tokio::time::delay_for(timeout_duration);
+            future::select(fut, timeout).await
+        }) {
+            future::Either::Left((Ok(resp), _)) => return resp,
+            future::Either::Left((Err(e), _)) => {
+                error!("kms request failed"; "error"=>?e);
+            }
+            future::Either::Right((_, _)) => {
+                error!("kms request timeout"; "timeout" => ?timeout_duration);
+            }
+        }
+    }
+    panic!("kms request failed in {} times", retry_limit)
+}
+
+pub struct KmsBackend {
+    file_backend: FileBackend,
+}
+
+impl KmsBackend {
+    pub fn new(config: KmsConfig, base: &str) -> Result<KmsBackend> {
+        let kms = AwsKms::new(config)?;
+        KmsBackend::with_kms(kms, base)
+    }
+
+    fn with_kms(kms: AwsKms, base: &str) -> Result<KmsBackend> {
+        let vendor_backend = WithMetadata {
+            // For now, we only support AWS.
+            metadata: vec![(MetadataKey::KmsVendor, AWS_KMS_VENDOR_NAME.to_vec())],
+            // ciphertext_key has already be ecrypted by KMS.
+            backend: PlainTextBackend::default(),
+        };
+
+        // Create base dir if it is missing.
+        create_dir_all(base)?;
+
+        // Read the master key or generate a new master key.
+        let key_path = Path::new(base).join(KMS_ENCRYPTION_KEY_NAME);
+
+        let mut runtime = Builder::new()
+            .basic_scheduler()
+            .thread_name("kms-pool")
+            .core_threads(1)
+            .enable_all()
+            .build()?;
+
+        let key = if !key_path.exists() {
+            let (ciphertext_key, plaintext_key) = kms.generate_data_key(&mut runtime)?;
+            let f = EncryptedFile::new(Path::new(base), KMS_ENCRYPTION_KEY_NAME);
+            f.write(&ciphertext_key, &vendor_backend)?;
+            plaintext_key
+        } else {
+            let f = EncryptedFile::new(Path::new(base), KMS_ENCRYPTION_KEY_NAME);
+            let ciphertext_key = f.read(&vendor_backend)?;
+            kms.decrypt(&mut runtime, &ciphertext_key)?
+        };
+
+        // Always use AES 256 for encrypting master key.
+        let method = KMS_DATA_KEY_METHOD;
+        let file_backend = FileBackend::new(method, key)?;
```

Comment:
The name of `FileBackend` is confusing, as it doesn't directly access a file, instead is back by an in-memory key. I would prefer making `FileBackend` handle file reading logic, and `KmsBackend` not depend on `FileBackend`.

Reply:
Only master key is generated from KMS, file dict and key dict still need to be encrypted into/ decrypted from files.

Reply:
I mean, right now `FileBackend` accepts a `Vec<u8>` as key, where it should accept a path of master key file. And I don't see why we need to use `FileBackend` inside `KmsBackend`.